### PR TITLE
Fixed wrong implementation of GenericOauth2Client

### PR DESCRIPTION
**Pull request by julioa**
I've found that previous implementation of GenericOAuth2Client I sent was adding additional attributes to the class instead of using the configuration object.

I've also made customParameters more flexible for different client implementations.

### CODE REVIEW

File: pac4j-oauth/src/main/java/org/pac4j/oauth/config/OAuth20Configuration.java [hunk 1]
```diff
@@ -32,14 +32,22 @@ public String getStateSessionAttributeName() {
...
         return customParams;
     }
 
-    public void setCustomParams(final Map<String, String> customParams) {
+    public Object getCustomParam(String name) {
+        return (customParams!=null)?customParams.get(name):null;
+    }
+
+    public void setCustomParams(final Map<String, Object> customParams) {
         this.customParams = customParams;
```
Comment (leleuj) [inline pac4j-oauth/src/main/java/org/pac4j/oauth/config/OAuth20Configuration.java]: I would add a `CommonHelper.assertNotNull` for the `customParams`.

**Comment by leleuj** [comment]
I like the idea. A few comments. Fix the build also.

**Comment by leleuj** [comment]
Can you sync with master? I think I have fixed the build error.

**Comment by julioa** [comment]
I still have a doubt for these changes: I'm not sure if mandatory config parameters for the clients should be declared in its constructor or taken from the configuration object. What policy are you using in others clients?

**Comment by leleuj** [comment]
A specific parameter for a client would be attached to the client (via its constructor), though custom parameters should be attached to the configuration.

File: pac4j-oauth/src/main/java/org/pac4j/oauth/redirect/OAuth20RedirectActionBuilder.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ public RedirectAction redirect(final WebContext context) throws HttpAction {
 
                 service = this.configuration.getService();
             }
-            final String authorizationUrl = service.getAuthorizationUrl(this.configuration.getCustomParams());
+            final String authorizationUrl = service.getAuthorizationUrl();
```
Comment (leleuj) [inline pac4j-oauth/src/main/java/org/pac4j/oauth/redirect/OAuth20RedirectActionBuilder.java]: One last question: why custom parameters are not passed anymore to the `getAuthorizationUrl()` method?

Comment (julioa) [inline pac4j-oauth/src/main/java/org/pac4j/oauth/redirect/OAuth20RedirectActionBuilder.java]: The parameter in getAuthorizationUrl are used to add additional parameters when building the authorization url by the scribe api implementation, as in org.pac4j.scribe.builder.api.GenericApi20, so customParameters should not be used here.

Comment (leleuj) [inline pac4j-oauth/src/main/java/org/pac4j/oauth/redirect/OAuth20RedirectActionBuilder.java]: OK. Let's merge it.

**Comment by leleuj** [comment]
Reviewing this once more, I don't think we took the right path. Custom parameters should be used for additional parameters on the authorization URL (like for OpenID Connect) and well-known parameters should be defined via setters/getters in the clients (to make the contract obvious).

I'm reverting this change. Would you mind revisiting it while keeping the setters/getters for the expected parameters?

**Comment by leleuj** [comment]
@julioa reverting with: https://github.com/pac4j/pac4j/pull/781

**Comment by leleuj** [comment]
It would be great to have a manual test as well (https://github.com/pac4j/pac4j/tree/master/pac4j-oauth/src/test/java/org/pac4j/oauth/run) and improve the documentation (https://github.com/pac4j/pac4j/blob/master/documentation/docs/clients/oauth.md).

**Pull request merged**